go/tests/googleai_test.go

```diff
@@ -0,0 +1,176 @@
+// limitations under the License.
+
+package tests
+
+import (
+	"cmp"
+	"context"
+	"flag"
+	"fmt"
+	"net/http"
+	"path/filepath"
+	"slices"
+	"testing"
+
+	"github.com/firebase/genkit/go/ai"
+	"github.com/firebase/genkit/go/core"
+	"github.com/firebase/genkit/go/core/tracing"
+	"github.com/firebase/genkit/go/internal"
+	"github.com/firebase/genkit/go/plugins/googleai"
+	gocmp "github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/jba/xltest/go/xltest"
+	"google.golang.org/api/option"
+)
+
+var update = flag.Bool("update", false, "update golden files with results")
+
+func TestGoogleAI(t *testing.T) {
+	ctx := context.Background()
+	err := googleai.Init(ctx, googleai.Config{
+		ClientOptions: []option.ClientOption{option.WithHTTPClient(mockClient())},
+		Models:        []string{"gemini-1.5-pro"},
+		Embedders:     []string{"embedding-001"},
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run("embed", func(t *testing.T) {
+		test, err := xltest.ReadFile(filepath.Join("testdata", "googleai-embedder.yaml"))
+		if err != nil {
+			t.Fatal(err)
+		}
+		valfunc := validateEmbedder
+		if *update {
+			valfunc = updateEmbedder
+		}
+		test.Run(t, func(text string) (embedderResult, error) {
+			ts := &testTraceStore{}
+			remove := core.SetDevTraceStore(ts)
+			defer remove()
+			vals, err := ai.Embed(ctx, googleai.Embedder("embedding-001"), &ai.EmbedRequest{Document: ai.DocumentFromText(text, nil)})
+			if err != nil {
+				return embedderResult{}, err
+			}
+			return embedderResult{vals, ts.td}, nil
+		}, valfunc)
+	})
+}
+
+type embedderResult struct {
+	values []float32
+	trace  *tracing.Data
+}
+
+// validateEmbedder compares the results from running an embedder with
+// the desired results. The latter are taken from a YAML file and so are in raw
+// unmarshalled form.
+func validateEmbedder(got embedderResult, rawWant map[string]any) error {
+	var want embedderResult
+	rawVals := rawWant["values"].([]any)
+	f32s := make([]float32, len(rawVals))
+	for i, v := range rawVals {
+		f32s[i] = float32(v.(float64))
+	}
+	want.values = f32s
+	traceFile := rawWant["traceFile"].(string)
+	if err := internal.ReadJSONFile(filepath.Join("testdata", traceFile), &want.trace); err != nil {
+		return err
+	}
+	return compareEmbedderResults(got, want)
+}
+
+func compareEmbedderResults(got, want embedderResult) error {
+	if !slices.Equal(got.values, want.values) {
+		return fmt.Errorf("values: got %v, want %v", got.values, want.values)
+	}
+	renameSpans(got.trace)
+	renameSpans(want.trace)
+	opts := []gocmp.Option{
+		cmpopts.IgnoreFields(tracing.Data{}, "TraceID", "StartTime", "EndTime"),
+		cmpopts.IgnoreFields(tracing.SpanData{}, "TraceID", "StartTime", "EndTime"),
+	}
+	if diff := gocmp.Diff(want.trace, got.trace, opts...); diff != "" {
+		return fmt.Errorf("traces: %s", diff)
+	}
+	return nil
+}
+
+// renameSpans changes the keys of td.Spans to s0, s1, ... in order of the span start time,
```

Comment:
Pedantically I think it's s000, s001, ....

Reply:
done